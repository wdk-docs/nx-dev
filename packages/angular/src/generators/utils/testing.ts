import type { Tree } from '@nrwl/devkit';
import { wrapAngularDevkitSchematic } from '@nrwl/devkit/ngcli-adapter';
import { createTreeWithEmptyWorkspace } from '@nrwl/devkit/testing';
import { Linter } from '@nrwl/linter';
import { UnitTestRunner } from '../../utils/test-runners';
import libraryGenerator from '../library/library';

export async function createStorybookTestWorkspaceForLib(
  libName: string
): Promise<Tree> {
  let tree = createTreeWithEmptyWorkspace();

  const moduleGenerator = wrapAngularDevkitSchematic(
    '@schematics/angular',
    'module'
  );
  const componentGenerator = wrapAngularDevkitSchematic(
    '@schematics/angular',
    'component'
  );

  await libraryGenerator(tree, {
    name: libName,
    buildable: false,
    linter: Linter.EsLint,
    publishable: false,
    simpleModuleName: false,
    skipFormat: false,
    unitTestRunner: UnitTestRunner.Jest,
  });

  await componentGenerator(tree, {
    name: 'test-button',
    project: libName,
  });

  tree.write(
    `libs/${libName}/src/lib/test-button/test-button.component.ts`,
    `import { Component, Input } from '@angular/core';

export type ButtonStyle = 'default' | 'primary' | 'accent';

@Component({
  selector: 'proj-test-button',
  templateUrl: './test-button.component.html',
  styleUrls: ['./test-button.component.css']
})
export class TestButtonComponent {
  @Input('buttonType') type = 'button';
  @Input() style: ButtonStyle = 'default';
  @Input() age?: number;
  @Input() isOn = false;
}`
  );

  tree.write(
    `libs/${libName}/src/lib/test-button/test-button.component.html`,
    `<button [attr.type]="type" [ngClass]="style"></button>`
  );

  const modulePath = `libs/${libName}/src/lib/${libName}.module.ts`;
  tree.write(
    modulePath,
    `import * as ButtonExports from './test-button/test-button.component';
    ${tree.read(modulePath)}`
  );

  // create a module with component that gets exported in a barrel file
  await moduleGenerator(tree, {
    name: 'barrel',
    project: libName,
  });

  await componentGenerator(tree, {
    name: 'barrel-button',
    project: libName,
    path: `libs/${libName}/src/lib/barrel`,
    module: 'barrel',
  });

  tree.write(
    `libs/${libName}/src/lib/barrel/barrel-button/index.ts`,
    `export * from './barrel-button.component';`
  );

  tree.write(
    `libs/${libName}/src/lib/barrel/barrel.module.ts`,
    `import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { BarrelButtonComponent } from './barrel-button';

@NgModule({
  imports: [CommonModule],
  declarations: [BarrelButtonComponent],
})
export class BarrelModule {}`
  );

  // create a module with components that get Angular exported and declared by variable
  await moduleGenerator(tree, {
    name: 'variable-declare',
    project: libName,
  });

  await componentGenerator(tree, {
    name: 'variable-declare-button',
    project: libName,
    path: `libs/${libName}/src/lib/variable-declare`,
    module: 'variable-declare',
  });

  await componentGenerator(tree, {
    name: 'variable-declare-view',
    project: libName,
    path: `libs/${libName}/src/lib/variable-declare`,
    module: 'variable-declare',
  });

  tree.write(
    `libs/${libName}/src/lib/variable-declare/variable-declare.module.ts`,
    `import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { VariableDeclareButtonComponent } from './variable-declare-button/variable-declare-button.component';
import { VariableDeclareViewComponent } from './variable-declare-view/variable-declare-view.component';

const COMPONENTS = [
  VariableDeclareButtonComponent,
  VariableDeclareViewComponent
]

@NgModule({
  imports: [CommonModule],
  declarations: COMPONENTS,
  exports: COMPONENTS
})
export class VariableDeclareModule {}`
  );

  // create a module with components that get Angular exported and declared by variable
  await moduleGenerator(tree, {
    name: 'variable-spread-declare',
    project: libName,
  });

  await componentGenerator(tree, {
    name: 'variable-spread-declare-button',
    project: libName,
    path: `libs/${libName}/src/lib/variable-spread-declare`,
    module: 'variable-spread-declare',
  });

  await componentGenerator(tree, {
    name: 'variable-spread-declare-view',
    project: libName,
    path: `libs/${libName}/src/lib/variable-spread-declare`,
    module: 'variable-spread-declare',
  });

  await componentGenerator(tree, {
    name: 'variable-spread-declare-anotherview',
    project: libName,
    path: `libs/${libName}/src/lib/variable-spread-declare`,
    module: 'variable-spread-declare',
  });

  tree.write(
    `libs/${libName}/src/lib/variable-spread-declare/variable-spread-declare.module.ts`,
    `import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { VariableSpreadDeclareButtonComponent } from './variable-spread-declare-button/variable-spread-declare-button.component';
import { VariableSpreadDeclareViewComponent } from './variable-spread-declare-view/variable-spread-declare-view.component';
import { VariableSpreadDeclareAnotherviewComponent } from './variable-spread-declare-anotherview/variable-spread-declare-anotherview.component';

const COMPONENTS = [ 
  VariableSpreadDeclareButtonComponent, 
  VariableSpreadDeclareViewComponent 
]

@NgModule({
  imports: [CommonModule],
  declarations: [...COMPONENTS, VariableSpreadDeclareAnotherviewComponent],
})
export class VariableSpreadDeclareModule {}`
  );

  // create a module where declared components are pulled from a static member of the module
  await moduleGenerator(tree, {
    name: 'static-member-declarations',
    project: libName,
  });

  await componentGenerator(tree, {
    name: 'cmp1',
    project: libName,
    path: `libs/${libName}/src/lib/static-member-declarations`,
    module: 'static-member-declarations',
  });

  await componentGenerator(tree, {
    name: 'cmp2',
    project: libName,
    path: `libs/${libName}/src/lib/static-member-declarations`,
    module: 'static-member-declarations',
  });

  tree.write(
    `libs/${libName}/src/lib/static-member-declarations/static-member-declarations.module.ts`,
    `import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { Cmp1Component } from './cmp1/cmp1.component';
import { Cmp2Component } from './cmp2/cmp2.component';

@NgModule({
  imports: [CommonModule],
  declarations: StaticMemberDeclarationsModule.COMPONENTS,
  exports: StaticMemberDeclarationsModule.COMPONENTS
})
export class StaticMemberDeclarationsModule {
  static readonly COMPONENTS = [Cmp1Component, Cmp2Component];
}`
  );

  // create another button in a nested subpath
  await moduleGenerator(tree, {
    name: 'nested',
    project: libName,
    path: `libs/${libName}/src/lib`,
  });

  await componentGenerator(tree, {
    name: 'nested-button',
    project: libName,
    module: 'nested',
    path: `libs/${libName}/src/lib/nested`,
  });

  await componentGenerator(tree, {
    name: 'test-other',
    project: libName,
  });

  return tree;
}
