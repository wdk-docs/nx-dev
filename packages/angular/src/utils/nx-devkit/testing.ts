import type { Tree } from '@nrwl/devkit';
import { names, updateJson } from '@nrwl/devkit';

export interface AppConfig {
  appName: string; // name of app
  appModule: string; // app/app.module.ts in the above sourceDir
}
export interface LibConfig {
  name: string;
  module: string;
  barrel: string;
}

var appConfig: AppConfig; // configure built in createApp()
var libConfig: LibConfig;

export function getAppConfig(): AppConfig {
  return appConfig;
}
export function getLibConfig(): LibConfig {
  return libConfig;
}

export function createApp(
  tree: Tree,
  appName: string,
  routing: boolean = true
) {
  appName = names(appName).fileName;
  // save for getAppDir() lookup by external *.spec.ts tests
  appConfig = {
    appName,
    appModule: `/apps/${appName}/src/app/app.module.ts`,
  };

  tree.write(
    appConfig.appModule,
    `
     import { NgModule } from '@angular/core';
     import { BrowserModule } from '@angular/platform-browser';
     ${routing ? "import { RouterModule } from '@angular/router'" : ''};
     import { AppComponent } from './app.component';
     @NgModule({
       imports: [BrowserModule, ${routing ? 'RouterModule.forRoot([])' : ''}],
       declarations: [AppComponent],
       bootstrap: [AppComponent]
     })
     export class AppModule {}
  `
  );
  tree.write(
    `/apps/${appName}/src/main.ts`,
    `
    import { enableProdMode } from '@angular/core';
    import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';

    import { AppModule } from './app/app.module';
    import { environment } from './environments/environment';

    if (environment.production) {
      enableProdMode();
    }

    platformBrowserDynamic()
      .bootstrapModule(AppModule)
      .catch(err => console.log(err));
  `
  );
  tree.write(
    `/apps/${appName}/tsconfig.app.json`,
    JSON.stringify({
      include: ['**/*.ts'],
    })
  );
  tree.write(
    `/apps/${appName}-e2e/tsconfig.e2e.json`,
    JSON.stringify({
      include: ['../**/*.ts'],
    })
  );
  tree.write(
    '/workspace.json',
    JSON.stringify({
      newProjectRoot: '',
      version: 1,
      projects: {
        [appName]: {
          root: `apps/${appName}`,
          sourceRoot: `apps/${appName}/src`,
          architect: {
            build: {
              options: {
                main: `apps/${appName}/src/main.ts`,
              },
            },
            serve: {
              options: {},
            },
          },
          tags: [],
        },
      },
    })
  );
}

export function createLib(tree: Tree, libName: string) {
  const { name, className, fileName, propertyName } = names(libName);

  libConfig = {
    name,
    module: `/libs/${propertyName}/src/lib/${fileName}.module.ts`,
    barrel: `/libs/${propertyName}/src/index.ts`,
  };

  tree.write(
    libConfig.module,
    `
      import { NgModule } from '@angular/core';
      import { CommonModule } from '@angular/common';
      @NgModule({
        imports: [
          CommonModule
        ],
        providers: []
      })
      export class ${className}Module { }
  `
  );
  tree.write(
    libConfig.barrel,
    `
    export * from './lib/${fileName}.module';
  `
  );
}
