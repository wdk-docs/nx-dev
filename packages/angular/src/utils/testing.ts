import { join } from 'path';
import { SchematicTestRunner } from '@angular-devkit/schematics/testing';
import { Rule, Tree } from '@angular-devkit/schematics';

const testRunner = new SchematicTestRunner(
  '@nrwl/angular',
  join(__dirname, '../../generators.json')
);

testRunner.registerCollection(
  '@nrwl/jest',
  join(__dirname, '../../../jest/generators.json')
);

testRunner.registerCollection(
  '@nrwl/workspace',
  join(__dirname, '../../../workspace/generators.json')
);

testRunner.registerCollection(
  '@nrwl/cypress',
  join(__dirname, '../../../cypress/generators.json')
);

testRunner.registerCollection(
  '@nrwl/storybook',
  join(__dirname, '../../../storybook/generators.json')
);

const migrationTestRunner = new SchematicTestRunner(
  '@nrwl/workspace',
  join(__dirname, '../../migrations.json')
);

export function runMigration<SchemaOptions = any>(
  schematicName: string,
  options: SchemaOptions,
  tree: Tree
) {
  return migrationTestRunner
    .runSchematicAsync(schematicName, options, tree)
    .toPromise();
}

export function callRule(rule: Rule, tree: Tree) {
  return testRunner.callRule(rule, tree).toPromise();
}
