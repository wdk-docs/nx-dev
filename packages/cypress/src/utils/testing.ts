import { join } from 'path';
import { SchematicTestRunner } from '@angular-devkit/schematics/testing';
import { Tree, Rule } from '@angular-devkit/schematics';

const testRunner = new SchematicTestRunner(
  '@nrwl/cypress',
  join(__dirname, '../../generators.json')
);

const migrationRunner = new SchematicTestRunner(
  '@nrwl/cypress/migrations',
  join(__dirname, '../../migrations.json')
);

export function runSchematic(schematicName: string, options: any, tree: Tree) {
  return testRunner.runSchematicAsync(schematicName, options, tree).toPromise();
}

export function callRule(rule: Rule, tree: Tree) {
  return testRunner.callRule(rule, tree).toPromise();
}

export function runMigration(migrationName: string, options: any, tree: Tree) {
  return migrationRunner
    .runSchematicAsync(migrationName, options, tree)
    .toPromise();
}
