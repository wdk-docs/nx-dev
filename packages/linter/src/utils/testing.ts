import { SchematicTestRunner } from '@angular-devkit/schematics/testing';
import { join } from 'path';
import { Tree } from '@angular-devkit/schematics';

const migrationTestRunner = new SchematicTestRunner(
  '@nrwl/linter-migrations',
  join(__dirname, '../../migrations.json')
);

export function runMigration<T = any>(
  migrationName: string,
  options: T,
  tree: Tree
) {
  return migrationTestRunner
    .runSchematicAsync<T>(migrationName, options, tree)
    .toPromise();
}
