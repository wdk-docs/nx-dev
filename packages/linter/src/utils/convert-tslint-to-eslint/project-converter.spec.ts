import {
  addDependenciesToPackageJson,
  addProjectConfiguration,
  NxJsonConfiguration,
  readJson,
  readProjectConfiguration,
  Tree,
  writeJson,
} from '@nrwl/devkit';
import { createTreeWithEmptyWorkspace } from '@nrwl/devkit/testing';
import { ProjectConverter } from './project-converter';

/**
 * Don't run actual child_process implementation of installPackagesTask()
 */
jest.mock('child_process');

/**
 * Don't run the conversion util, it touches the file system and has its own tests
 */
jest.mock('./utils', () => {
  return {
    // Since upgrading to (ts-)jest 26 this usage of this mock has caused issues...
    // @ts-ignore
    ...jest.requireActual<any>('./utils'),
    convertTSLintConfig: jest.fn(() => {
      return {
        convertedESLintConfig: {
          rules: {
            'some-converted-rule': 'error',
          },
        },
        unconvertedTSLintRules: [],
        ensureESLintPlugins: [],
      };
    }),
  };
});

describe('ProjectConverter', () => {
  let host: Tree;
  const projectName = 'foo';
  const projectRoot = `apps/${projectName}`;

  beforeEach(async () => {
    // Clean up any previous dry run simulations
    process.argv = process.argv.filter(
      (a) => !['--dry-run', '--dryRun', '-d'].includes(a)
    );
    host = createTreeWithEmptyWorkspace();
    addProjectConfiguration(host, projectName, {
      root: projectRoot,
      projectType: 'application',
      targets: {
        /**
         * LINT TARGET CONFIG - BEFORE CONVERSION
         *
         * TSLint executor configured for the project
         */
        lint: {
          executor: '@angular-devkit/build-angular:tslint',
          options: {
            exclude: ['**/node_modules/**', `!${projectRoot}/**/*`],
            tsConfig: [`${projectRoot}/tsconfig.app.json`],
          },
        },
      },
      /**
       * PROJECT-LEVEL GENERATOR CONFIG - BEFORE CONVERSION
       *
       * Default set to tslint, using shorthand syntax
       */
      generators: {
        '@nrwl/angular:library': {
          linter: 'tslint',
        },
        '@nrwl/angular:application': {
          e2eTestRunner: 'cypress',
          linter: 'eslint',
          unitTestRunner: 'jest',
        },
      },
    });
  });

  it('should throw if --dry-run is set', () => {
    writeJson(host, 'tslint.json', {});
    writeJson(host, `${projectRoot}/tslint.json`, {});

    process.argv.push('--dry-run');

    expect(
      () =>
        new ProjectConverter({
          host,
          projectName,
          ignoreExistingTslintConfig: false,
          eslintInitializer: () => undefined,
        })
    ).toThrowErrorMatchingSnapshot();
  });

  it('should throw if --dryRun is set', () => {
    writeJson(host, 'tslint.json', {});
    writeJson(host, `${projectRoot}/tslint.json`, {});

    process.argv.push('--dryRun');

    expect(
      () =>
        new ProjectConverter({
          host,
          projectName,
          ignoreExistingTslintConfig: false,
          eslintInitializer: () => undefined,
        })
    ).toThrowErrorMatchingSnapshot();
  });

  it('should throw if -d is set', () => {
    writeJson(host, 'tslint.json', {});
    writeJson(host, `${projectRoot}/tslint.json`, {});

    process.argv.push('-d');

    expect(
      () =>
        new ProjectConverter({
          host,
          projectName,
          ignoreExistingTslintConfig: false,
          eslintInitializer: () => undefined,
        })
    ).toThrowErrorMatchingSnapshot();
  });

  it('should throw if no root tslint.json is found', () => {
    expect(
      () =>
        new ProjectConverter({
          host,
          projectName,
          ignoreExistingTslintConfig: false,
          eslintInitializer: () => undefined,
        })
    ).toThrowErrorMatchingSnapshot();
  });

  it('should throw if no project tslint.json is found', () => {
    writeJson(host, 'tslint.json', {});

    expect(
      () =>
        new ProjectConverter({
          host,
          projectName,
          ignoreExistingTslintConfig: false,
          eslintInitializer: () => undefined,
        })
    ).toThrowErrorMatchingSnapshot();
  });

  it('should not throw when not in dry-run and config files successfully found', () => {
    writeJson(host, 'tslint.json', {});
    writeJson(host, `${projectRoot}/tslint.json`, {});

    expect(
      () =>
        new ProjectConverter({
          host,
          projectName,
          ignoreExistingTslintConfig: false,
          eslintInitializer: () => undefined,
        })
    ).not.toThrow();
  });

  describe('ignoreExistingTslintConfig', () => {
    describe('ignoreExistingTslintConfig: false', () => {
      it('should use existing TSLint configs and merge their converted ESLint equivalents with recommended ESLint configs', async () => {
        const { convertTSLintConfig } = require('./utils');
        (convertTSLintConfig as jest.Mock).mockClear();

        writeJson(host, 'tslint.json', {});
        writeJson(host, `${projectRoot}/tslint.json`, {});

        const projectConverterDiscardFalse = new ProjectConverter({
          host,
          projectName,
          ignoreExistingTslintConfig: false,
          eslintInitializer: () => {
            writeJson(host, '.eslintrc.json', {
              rules: {
                'some-recommended-root-eslint-rule': 'error',
              },
            });
            writeJson(host, `${projectRoot}/.eslintrc.json`, {
              rules: {
                'some-recommended-project-eslint-rule': 'error',
              },
            });
            return Promise.resolve(undefined);
          },
        });
        await projectConverterDiscardFalse.initESLint();

        // Existing ESLint rules from init stage will be merged with converted rules
        await projectConverterDiscardFalse.convertRootTSLintConfig((j) => j);
        expect(readJson(host, '.eslintrc.json')).toMatchSnapshot();

        // Existing ESLint rules from init stage will be merged with converted rules
        await projectConverterDiscardFalse.convertProjectConfig((j) => j);
        expect(
          readJson(host, `${projectRoot}/.eslintrc.json`)
        ).toMatchSnapshot();

        expect(convertTSLintConfig).toHaveBeenCalledTimes(2);
      });
    });

    describe('ignoreExistingTslintConfig: true', () => {
      it('should ignore existing TSLint configs and just reset the project to use recommended ESLint configs', async () => {
        const { convertTSLintConfig } = require('./utils');
        (convertTSLintConfig as jest.Mock).mockClear();

        writeJson(host, 'tslint.json', {});
        writeJson(host, `${projectRoot}/tslint.json`, {});

        const projectConverterDiscardTrue = new ProjectConverter({
          host,
          projectName,
          ignoreExistingTslintConfig: true,
          eslintInitializer: () => {
            writeJson(host, '.eslintrc.json', {
              rules: {
                'some-recommended-root-eslint-rule': 'error',
              },
            });
            writeJson(host, `${projectRoot}/.eslintrc.json`, {
              rules: {
                'some-recommended-project-eslint-rule': 'error',
              },
            });
            return Promise.resolve(undefined);
          },
        });
        await projectConverterDiscardTrue.initESLint();

        // Should not contain any converted rules, just existing ESLint rules from init stage
        await projectConverterDiscardTrue.convertRootTSLintConfig((j) => j);
        expect(readJson(host, '.eslintrc.json')).toMatchSnapshot();

        // Should not contain any converted rules, just existing ESLint rules from init stage
        await projectConverterDiscardTrue.convertProjectConfig((j) => j);
        expect(
          readJson(host, `${projectRoot}/.eslintrc.json`)
        ).toMatchSnapshot();

        expect(convertTSLintConfig).not.toHaveBeenCalled();
      });
    });
  });

  describe('setDefaults()', () => {
    it('should save in nx.json', async () => {
      writeJson(host, 'tslint.json', {});
      writeJson(host, `${projectRoot}/tslint.json`, {});

      const projectConverter = new ProjectConverter({
        host,
        projectName,
        ignoreExistingTslintConfig: false,
        eslintInitializer: () => undefined,
      });

      const nxJson = readJson<NxJsonConfiguration>(host, 'nx.json');
      nxJson.generators = {
        '@nrwl/angular': {
          application: {
            linter: 'tslint',
          },
          library: {
            linter: 'tslint',
          },
        },
      };
      writeJson(host, 'nx.json', nxJson);

      // BEFORE - no entry for convert-tslint-to-eslint wthin @nrwl/angular generators
      expect(readJson(host, 'nx.json')).toMatchSnapshot();

      projectConverter.setDefaults('@nrwl/angular', {
        ignoreExistingTslintConfig: true,
        removeTSLintIfNoMoreTSLintTargets: true,
      });

      // AFTER (1) - convert-tslint-to-eslint wthin @nrwl/angular generators has removeTSLintIfNoMoreTSLintTargets set to true
      expect(readJson(host, 'nx.json')).toMatchSnapshot();

      projectConverter.setDefaults('@nrwl/angular', {
        ignoreExistingTslintConfig: false,
        removeTSLintIfNoMoreTSLintTargets: false,
      });

      // AFTER (2) - convert-tslint-to-eslint wthin @nrwl/angular generators has removeTSLintIfNoMoreTSLintTargets set to false
      expect(readJson(host, 'nx.json')).toMatchSnapshot();
    });
  });

  describe('removeTSLintFromWorkspace()', () => {
    it('should remove all relevant traces of TSLint from the workspace', async () => {
      writeJson(host, 'tslint.json', {});
      writeJson(host, `${projectRoot}/tslint.json`, {});

      const projectConverter = new ProjectConverter({
        host,
        projectName,
        ignoreExistingTslintConfig: false,
        eslintInitializer: () => undefined,
      });

      await addDependenciesToPackageJson(
        host,
        {},
        {
          codelyzer: 'latest',
          tslint: 'latest',
        }
      )();

      const nxJson = readJson<NxJsonConfiguration>(host, 'nx.json');
      // Not using shorthand syntax this time
      nxJson.generators = {
        '@nrwl/angular': {
          application: {
            linter: 'tslint',
          },
          library: {
            linter: 'tslint',
          },
        },
      };
      writeJson(host, 'nx.json', nxJson);

      // BEFORE - tslint and codelyzer are present
      expect(readJson(host, 'package.json')).toMatchSnapshot();

      // BEFORE - tslint set as both global linter for @nrwl/angular generators
      expect(readJson(host, 'nx.json')).toMatchSnapshot();

      expect(readProjectConfiguration(host, projectName)).toMatchSnapshot();

      await projectConverter.removeTSLintFromWorkspace()();

      // AFTER - it should remove tslint and codelyzer
      expect(readJson(host, 'package.json')).toMatchSnapshot();

      // AFTER - generators config from global project-level settings removed (because eslint is always default)
      expect(readJson(host, 'nx.json')).toMatchSnapshot();

      // AFTER - generators config from project-level settings removed (because eslint is always default)
      expect(readProjectConfiguration(host, projectName)).toMatchSnapshot();
    });

    it('should remove the entry in generators for convert-tslint-to-eslint because it is no longer needed', async () => {
      writeJson(host, 'tslint.json', {});
      writeJson(host, `${projectRoot}/tslint.json`, {});

      const projectConverter = new ProjectConverter({
        host,
        projectName,
        ignoreExistingTslintConfig: false,
        eslintInitializer: () => undefined,
      });

      const nxJson = readJson<NxJsonConfiguration>(host, 'nx.json');
      nxJson.generators = {
        '@nrwl/angular': {
          'convert-tslint-to-eslint': {
            removeTSLintIfNoMoreTSLintTargets: true,
          },
        },
      };
      writeJson(host, 'nx.json', nxJson);

      // BEFORE - convert-tslint-to-eslint wthin @nrwl/angular generators has a value for removeTSLintIfNoMoreTSLintTargets
      expect(readJson(host, 'nx.json')).toMatchSnapshot();

      await projectConverter.removeTSLintFromWorkspace()();

      // AFTER - generators config no longer has a reference to convert-tslint-to-eslint
      expect(readJson(host, 'nx.json')).toMatchSnapshot();
    });
  });
});
