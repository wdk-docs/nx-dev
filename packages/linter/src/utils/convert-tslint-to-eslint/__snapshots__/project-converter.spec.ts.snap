// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectConverter ignoreExistingTslintConfig ignoreExistingTslintConfig: false should use existing TSLint configs and merge their converted ESLint equivalents with recommended ESLint configs 1`] = `
Object {
  "overrides": Array [
    Object {
      "files": Array [
        "*.ts",
      ],
      "rules": Object {
        "some-converted-rule": "error",
      },
    },
  ],
  "rules": Object {
    "some-recommended-root-eslint-rule": "error",
  },
}
`;

exports[`ProjectConverter ignoreExistingTslintConfig ignoreExistingTslintConfig: false should use existing TSLint configs and merge their converted ESLint equivalents with recommended ESLint configs 2`] = `
Object {
  "rules": Object {
    "some-converted-rule": "error",
    "some-recommended-project-eslint-rule": "error",
  },
}
`;

exports[`ProjectConverter ignoreExistingTslintConfig ignoreExistingTslintConfig: true should ignore existing TSLint configs and just reset the project to use recommended ESLint configs 1`] = `
Object {
  "rules": Object {
    "some-recommended-root-eslint-rule": "error",
  },
}
`;

exports[`ProjectConverter ignoreExistingTslintConfig ignoreExistingTslintConfig: true should ignore existing TSLint configs and just reset the project to use recommended ESLint configs 2`] = `
Object {
  "rules": Object {
    "some-recommended-project-eslint-rule": "error",
  },
}
`;

exports[`ProjectConverter removeTSLintFromWorkspace() should remove all relevant traces of TSLint from the workspace 1`] = `
Object {
  "dependencies": Object {},
  "devDependencies": Object {
    "codelyzer": "latest",
    "tslint": "latest",
  },
  "name": "test-name",
}
`;

exports[`ProjectConverter removeTSLintFromWorkspace() should remove all relevant traces of TSLint from the workspace 2`] = `
Object {
  "affected": Object {
    "defaultBase": "main",
  },
  "generators": Object {
    "@nrwl/angular": Object {
      "application": Object {
        "linter": "tslint",
      },
      "library": Object {
        "linter": "tslint",
      },
    },
  },
  "npmScope": "proj",
  "tasksRunnerOptions": Object {
    "default": Object {
      "options": Object {
        "cacheableOperations": Array [
          "build",
          "lint",
          "test",
          "e2e",
        ],
      },
      "runner": "@nrwl/workspace/tasks-runners/default",
    },
  },
}
`;

exports[`ProjectConverter removeTSLintFromWorkspace() should remove all relevant traces of TSLint from the workspace 3`] = `
Object {
  "generators": Object {
    "@nrwl/angular:application": Object {
      "e2eTestRunner": "cypress",
      "linter": "eslint",
      "unitTestRunner": "jest",
    },
    "@nrwl/angular:library": Object {
      "linter": "tslint",
    },
  },
  "projectType": "application",
  "root": "apps/foo",
  "targets": Object {
    "lint": Object {
      "executor": "@angular-devkit/build-angular:tslint",
      "options": Object {
        "exclude": Array [
          "**/node_modules/**",
          "!apps/foo/**/*",
        ],
        "tsConfig": Array [
          "apps/foo/tsconfig.app.json",
        ],
      },
    },
  },
}
`;

exports[`ProjectConverter removeTSLintFromWorkspace() should remove all relevant traces of TSLint from the workspace 4`] = `
Object {
  "dependencies": Object {},
  "devDependencies": Object {},
  "name": "test-name",
}
`;

exports[`ProjectConverter removeTSLintFromWorkspace() should remove all relevant traces of TSLint from the workspace 5`] = `
Object {
  "affected": Object {
    "defaultBase": "main",
  },
  "npmScope": "proj",
  "tasksRunnerOptions": Object {
    "default": Object {
      "options": Object {
        "cacheableOperations": Array [
          "build",
          "lint",
          "test",
          "e2e",
        ],
      },
      "runner": "@nrwl/workspace/tasks-runners/default",
    },
  },
}
`;

exports[`ProjectConverter removeTSLintFromWorkspace() should remove all relevant traces of TSLint from the workspace 6`] = `
Object {
  "generators": Object {
    "@nrwl/angular:application": Object {
      "e2eTestRunner": "cypress",
      "unitTestRunner": "jest",
    },
  },
  "projectType": "application",
  "root": "apps/foo",
  "targets": Object {
    "lint": Object {
      "executor": "@angular-devkit/build-angular:tslint",
      "options": Object {
        "exclude": Array [
          "**/node_modules/**",
          "!apps/foo/**/*",
        ],
        "tsConfig": Array [
          "apps/foo/tsconfig.app.json",
        ],
      },
    },
  },
}
`;

exports[`ProjectConverter removeTSLintFromWorkspace() should remove the entry in generators for convert-tslint-to-eslint because it is no longer needed 1`] = `
Object {
  "affected": Object {
    "defaultBase": "main",
  },
  "generators": Object {
    "@nrwl/angular": Object {
      "convert-tslint-to-eslint": Object {
        "removeTSLintIfNoMoreTSLintTargets": true,
      },
    },
  },
  "npmScope": "proj",
  "tasksRunnerOptions": Object {
    "default": Object {
      "options": Object {
        "cacheableOperations": Array [
          "build",
          "lint",
          "test",
          "e2e",
        ],
      },
      "runner": "@nrwl/workspace/tasks-runners/default",
    },
  },
}
`;

exports[`ProjectConverter removeTSLintFromWorkspace() should remove the entry in generators for convert-tslint-to-eslint because it is no longer needed 2`] = `
Object {
  "affected": Object {
    "defaultBase": "main",
  },
  "npmScope": "proj",
  "tasksRunnerOptions": Object {
    "default": Object {
      "options": Object {
        "cacheableOperations": Array [
          "build",
          "lint",
          "test",
          "e2e",
        ],
      },
      "runner": "@nrwl/workspace/tasks-runners/default",
    },
  },
}
`;

exports[`ProjectConverter setDefaults() should save in nx.json 1`] = `
Object {
  "affected": Object {
    "defaultBase": "main",
  },
  "generators": Object {
    "@nrwl/angular": Object {
      "application": Object {
        "linter": "tslint",
      },
      "library": Object {
        "linter": "tslint",
      },
    },
  },
  "npmScope": "proj",
  "tasksRunnerOptions": Object {
    "default": Object {
      "options": Object {
        "cacheableOperations": Array [
          "build",
          "lint",
          "test",
          "e2e",
        ],
      },
      "runner": "@nrwl/workspace/tasks-runners/default",
    },
  },
}
`;

exports[`ProjectConverter setDefaults() should save in nx.json 2`] = `
Object {
  "affected": Object {
    "defaultBase": "main",
  },
  "generators": Object {
    "@nrwl/angular": Object {
      "application": Object {
        "linter": "tslint",
      },
      "convert-tslint-to-eslint": Object {
        "ignoreExistingTslintConfig": true,
        "removeTSLintIfNoMoreTSLintTargets": true,
      },
      "library": Object {
        "linter": "tslint",
      },
    },
  },
  "npmScope": "proj",
  "tasksRunnerOptions": Object {
    "default": Object {
      "options": Object {
        "cacheableOperations": Array [
          "build",
          "lint",
          "test",
          "e2e",
        ],
      },
      "runner": "@nrwl/workspace/tasks-runners/default",
    },
  },
}
`;

exports[`ProjectConverter setDefaults() should save in nx.json 3`] = `
Object {
  "affected": Object {
    "defaultBase": "main",
  },
  "generators": Object {
    "@nrwl/angular": Object {
      "application": Object {
        "linter": "tslint",
      },
      "convert-tslint-to-eslint": Object {
        "ignoreExistingTslintConfig": false,
        "removeTSLintIfNoMoreTSLintTargets": false,
      },
      "library": Object {
        "linter": "tslint",
      },
    },
  },
  "npmScope": "proj",
  "tasksRunnerOptions": Object {
    "default": Object {
      "options": Object {
        "cacheableOperations": Array [
          "build",
          "lint",
          "test",
          "e2e",
        ],
      },
      "runner": "@nrwl/workspace/tasks-runners/default",
    },
  },
}
`;

exports[`ProjectConverter should throw if --dry-run is set 1`] = `
"NOTE: This generator does not support --dry-run. If you are running this in Nx Console, it should execute fine once you hit the \\"Run\\" button.
"
`;

exports[`ProjectConverter should throw if --dryRun is set 1`] = `
"NOTE: This generator does not support --dry-run. If you are running this in Nx Console, it should execute fine once you hit the \\"Run\\" button.
"
`;

exports[`ProjectConverter should throw if -d is set 1`] = `
"NOTE: This generator does not support --dry-run. If you are running this in Nx Console, it should execute fine once you hit the \\"Run\\" button.
"
`;

exports[`ProjectConverter should throw if no project tslint.json is found 1`] = `"We could not find a tslint.json for the selected project \\"apps/foo/tslint.json\\", maybe you have already migrated to ESLint?"`;

exports[`ProjectConverter should throw if no root tslint.json is found 1`] = `"We could not find a tslint.json at the root of your workspace, maybe you have already migrated to ESLint?"`;
